tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/caches/Cache.scala

```diff
@@ -0,0 +1,177 @@
+    } else {
+      // introduce a new constant
+      val (newContext, targetValue) = create(context, srcValue)
+      addToCache(srcValue, targetValue)
+      (newContext, targetValue)
+    }
+
+  /**
+   * Add a value to the cache.
+   *
+   * @param source
+   *   the source value
+   * @param target
+   *   the cached target value
+   */
+  protected def addToCache(source: SourceT, target: TargetT): Unit = {
+    _cache += source -> (target, level)
+    _reverseCache += target -> (source, level)
+  }
+
+  /**
+   * Get a previously cached value for a given source value, if there is one. Otherwise, return none.
+   *
+   * @param srcValue
+   *   a source value
+   * @return
+   *   Some(result), or None
+   */
+  def get(srcValue: SourceT): Option[TargetT] = cache.get(srcValue).map(_._1)
+
+  /**
+   * Find the key that was used to create a given value.
+   * @param value
+   *   a value, for which the key should be found
+   * @return
+   *   the key, if exists
+   */
+  def findKey(value: TargetT): Option[SourceT] = reverseCache.get(value).map(_._1)
+
+  /**
+   * Get the current level, that is the difference between the number of pushes and pops made so far.
+   *
+   * @return
+   *   the current level, always non-negative.
+   */
+  override def level: Int = _level
+
+  /**
+   * Save the current state and push it on the stack for a later recovery with pop. Increases level by 1.
+   */
+  override def snapshot(): Unit = _level += 1
+
+  /**
+   * Discard all entries added at the current level. Decreases level by 1.
+   *
+   * Importantly, pop may be called multiple times and thus it is not sufficient to save only the latest state.
+   */
+  override def revert(): Unit = revert(1)
+
+  /**
+   * Pop n times.
+   * @param n
+   *   pop n times, if n > 0, otherwise, do nothing
+   */
+  override def revert(n: Int): Unit = {
+    require(level >= n, s"Can't pop $n levels from a cache of level $level.")
+    _level -= n
+
+    def isOld[A, B](mapEntry: (A, (B, Int))): Boolean =
+      mapEntry._2._2 <= level
+
+    _cache = cache.filter(isOld)
+    _reverseCache = reverseCache.filter(isOld)
+  }
+
+  /**
+   * Clean the state completely.
+   */
+  override def dispose(): Unit = {
+    _cache = HashMap.empty
+    _reverseCache = HashMap.empty
+    _level = 0
+  }
+
+  /** SMT constraints */
+
+  /** Extra method, may add constraints for a single key-value pair */
+  def addConstraintsForKV(ctx: SolverContext)(k: SourceT, v: TargetT): Unit
+
+  /**
+   * Add implementation-specific constraints for all entries added at `lvl` or later.
+   */
+  def addConstraintsFromLevel(ctx: SolverContext)(lvl: Int): Unit
+
+  /**
+   * Add constraints for all entries added at any level.
+   */
+  def addConstraints(ctx: SolverContext): Unit = addConstraintsFromLevel(ctx)(0)
+
+  /**
+   * Add constraints for all entries added at the current level.
+   */
+  def addConstraintsForCurrentLevel(ctx: SolverContext): Unit = addConstraintsFromLevel(ctx)(level)
```

Comment:
I don't think these should go into an abstract cache implementation.

Introduce a separate trait?

Reply:
I agree in principle, but I was specifically asked to move these methods into `Cache` in https://github.com/informalsystems/apalache/pull/2648#discussion_r1264830100

> It seems very special purpose given the last three methods. Perhaps we can do without the trait for now and just stick with the abstract class?

I'll move them back out

Reply:
I don't have a final name in mind, but to me there exists something like `SmtSyncable`, i.e. a trait, that is shared by all entities (not just caches, in principle), which "have to" update SMT constraints for every operation they perform (possibly delayed).

Reply:
Hmmm, these are completely independent. The rest of the cache doesn't interacte with them, and vice versa.

So from a cohesion POV, I think they belong in a separate type.

Reply:
I think it really is as simple as having a layer between `Stacklike` (`Snaphottable` here) and `Cache`, which governs the connection between stack-reversions and obligations to SMT

Reply:
Ah I see if these methods
Are going to
Be used by several different classes and they are, as Thomas points out, totally orthogonal to this chaching utility that makes sense. Thanks for explaining! Sounds like they don’t belong in the stacklike trait either then? That was my particular concern before, but having them in a separate trait entirely for these reasons makes sense to me!

Reply:
I mean, classes/traits are ultimately just a bookkeeping tool, if we'd rather have 1 trait 2ith 6 methods over 2 with 3 each that's not really that bad. We're not building a general-purpose library here.

Reply:
> I mean, classes/traits are ultimately just a bookkeeping tool, if we'd rather have 1 trait with 6 methods over 2 with 3 each that's not really that bad. We're not building a general-purpose library here.

Yeah, but then it should probably be `SmtSomethingCache`, not just `Cache` 😄

Reply:
>afaict it's only going to be used by this one class. 

Transition exploration is also stacklike.

Reply:
> So should I just get rid of the entire `Stacklike` trait and just have push/pop on the Cache or what? And then an extension of that `Cache` with the SMT stuff?

I think you have a better sense of what the final design / class hierarchy will look like.
What do you propose?

Reply:
Well, as it stands, there'd be at most 2 broad categories of classes, which have reversals as part of their behavior, the caches and the transition explorer (and the former only need this behavior because of the latter (in the current implementation, there are 2 traits, `StackableContext` and `Recoverable[T]`).

I actually think it makes sense to have the stack-like interface shared; at the very least the concept of push and pop. If we're not doing multicore, we don't need snapshot/recover (from `Recoverable`). So I'd keep it as a simple interface, but the two categories obviously have different implementations of push/pop, and none of the implementations actually get reused.

Reply:
> but the two categories obviously have different implementations of push/pop

Yeah, this bothers me... Why should they refine the same type then? What's the shared interface, other than that they share the same signature?

Reply:
I mean, isn't that enough? A stack and a queue share the same push/pop interface, but they obviously have different implementations.

Reply:
Not if you ask Liskov and Wing. One is LIFO, the other FIFO.

Reply:
Actually, stack and queue comes up as the first example, if you look for behavioral subtyping: https://en.wikipedia.org/wiki/Behavioral_subtyping

Reply:
That post specifically says that it's dangerous to have a stack be a subclass of Queue, not that they shouldn't both inherit from an interface with push/pop

Reply:
You answered my question above about their shared specification

> What's the shared interface, other than that they have the same type signature?

with another question, which is answered in the post: 🤷 

> It is important to stress that whether a type S is a behavioral subtype of a type T depends only on the specification (i.e. the documentation) of type T; the implementation of type T, if it has any, is completely irrelevant to this question.

Reply:
Roll it back a bit, because I'm not sure I understand what concern you're pointing out. Am I correct in interpreting that you don't want `Cache` and also `TrEx` in the future both extending `Stacklike`? Because `Stacklike`'s behavior is too vague?

Reply:
Yes. There's no code sharing, because all defs in `Stacklike` are abstract. And the interface is vague, since it lacks a signature for adding elements. So I'd rather have the concrete defs of `push` and `pop` first defined in two places, with very concrete behavioral descriptions.

My real pet peeve here is the following: Let's say I Cmd-click on `push()` in this line:

https://github.com/informalsystems/apalache/blob/bd215d798f14befd05c3ec72dc66770cc76f0cc7/tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/SymbStateRewriterImpl.scala#L526

IntelliJ takes me into the `StackableContext` trait. Not good, I wanted to see the actual implementation. If I try to reverse-lookup implementations of `push()`, there's 9 to choose from, from caches to rewriters, to the actual solver context. For `Recoverable`, it's even worse, there's 11 implementations of `recover()` (including the actual TRex, caches, various execution contexts, rewriters, etc). And all I wanted is to follow as simple method call.

Reply:
Or, in shorter form: The only reason to have a `Stacklike` here, is if you actually had a `def doSomething(x: Stacklike)` that we wanted to call with either a `Cache` or a `TrEx`. IMO that's not the case here, so no reason to have a common subtype.

Reply:
> If I try to reverse-lookup implementations of push(), there's 9 to choose from, from caches to rewriters, to the actual solver context. For Recoverable, it's even worse, there's 11 implementations of recover()

Yeah, part of the problem is that there a few de facto chaches, that just don't extend `AbstractCache`, but reimplement push&pop in exactly the same way. Ideally there would be 2 sub-classes, one for caches and one for transitions.

> The only reason to have a Stacklike here, is if you actually had a def doSomething(x: Stacklike) that we wanted to call with either a Cache or a TrEx.

You know, I can get behind that